### PR TITLE
fix: compilation on clang-19

### PR DESCRIPTION
**Pull request by ruslan-ilesik**
## Code change checklist

- [X] I have ensured that all methods and functions are **fully documented** using doxygen style comments.
- [X] My code follows the [coding style guide](https://dpp.dev/coding-standards.html).
- [x] I tested that my change works before raising the PR.
- [x] I have ensured that I did not break any existing API calls.
- [X] I have not built my pull request using AI, a static analysis tool or similar without any human oversight.

Unable to test voice because of broken unit test for voice send. 

Replaced std::basic_string with std::vector

Fix for [issue 1129](https://github.com/brainboxdotcc/DPP/issues/1129)

### CODE REVIEW

**Comment by braindigitalis** [comment]
this compiles but have you checked that it works? seems like quite a big change.

**Comment by ruslan-ilesik** [comment]
> this compiles but have you checked that it works? seems like quite a big change.

No, I did not yet, I will try to run one of our examples bots tomorrow on this version of build

**Comment by ruslan-ilesik** [comment]
Okay, just checked it using 
https://dpp.dev/stream-mp3-discord-bot.html
Works fine.
Also ran code using valgrind, and played song multiple times

==19411== Process terminating with default action of signal 2 (SIGINT)
==19411==    at 0x67D7117: __futex_abstimed_wait_common64 (futex-internal.c:57)
==19411==    by 0x67D7117: __futex_abstimed_wait_common (futex-internal.c:87)
==19411==    by 0x67D7117: __futex_abstimed_wait_cancelable64 (futex-internal.c:139)
==19411==    by 0x67D9A40: __pthread_cond_wait_common (pthread_cond_wait.c:503)
==19411==    by 0x67D9A40: pthread_cond_wait@@GLIBC_2.3.2 (pthread_cond_wait.c:627)
==19411==    by 0x655390E: std::__1::condition_variable::wait(std::__1::unique_lock<std::__1::mutex>&) (in /usr/lib/llvm-19/lib/libc++.so.1.0)
==19411==    by 0x5AEFC70: dpp::cluster::start(bool)::$_1::operator()() const (in /usr/local/lib/libdpp.so.10.0.31)
==19411==    by 0x5AEFB8E: dpp::cluster::start(bool) (in /usr/local/lib/libdpp.so.10.0.31)
==19411==    by 0x112D6D: main (in /home/ilesik/test/a.out)
==19411==
==19411== HEAP SUMMARY:
==19411==     in use at exit: 18,821,403 bytes in 9,732 blocks
==19411==   total heap usage: 72,243 allocs, 62,511 frees, 98,628,767 bytes allocated
==19411==
==19411== LEAK SUMMARY:
==19411==    definitely lost: 0 bytes in 0 blocks
==19411==    indirectly lost: 0 bytes in 0 blocks
==19411==      possibly lost: 8,064 bytes in 19 blocks
==19411==    still reachable: 18,813,339 bytes in 9,713 blocks
==19411==         suppressed: 0 bytes in 0 blocks
==19411== Rerun with --leak-check=full to see details of leaked memory
==19411==
==19411== For lists of detected and suppressed errors, rerun with: -s
==19411== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)

**Comment by braindigitalis** [comment]
SIGINT means ctrl c

**Review by braindigitalis — CHANGES_REQUESTED** [review]
not sure this is right please check it works

File: src/dpp/discordvoiceclient.cpp [hunk 1]
```diff
@@ -709,22 +709,21 @@ void discord_voice_client::read_ready()
...
 		}
 
 		/* It's a "silence packet" - throw it away. */
-		if (packet.size() < 44) {
+		if (packet_size < 44) {
 			return;
 		}
 
-		if (uint8_t payload_type = packet[1] & 0b0111'1111;
+		if (uint8_t payload_type = buffer[1] & 0b0111'1111;
```
[the remark this answers is not in the document]
Comment (ruslan-ilesik) [reply src/dpp/discordvoiceclient.cpp:726]: I didn't rename it, packet variable does not exist anymore. It was basic_string_view created from buffer which is deprecated

File: src/dpp/discordvoiceclient.cpp [hunk 2]
```diff
@@ -783,14 +783,15 @@ void discord_voice_client::read_ready()
...
+                        decrypted_data += ext_header_len + ext_len;
+                        decrypted_data_len -= ext_header_len + ext_len;
 		}
 
 		/*
 		 * We're left with the decrypted, opus-encoded data.
 		 * Park the payload and decode on the voice courier thread.
 		 */
-		vp.vr->audio_data.assign(decrypted_data);
+		vp.vr->audio_data.assign(decrypted_data,decrypted_data + decrypted_data_len);
```
[the remark this answers is not in the document]
Comment (ruslan-ilesik) [reply src/dpp/discordvoiceclient.cpp:794]: https://en.cppreference.com/w/cpp/container/vector/assign

Replaces the contents with copies of those in the range [first, last). The behavior is undefined if either argument is an iterator into *this.
This overload has the same effect as overload (1) if InputIt is an integral type.

(until C++11)
This overload participates in overload resolution only if InputIt satisfies LegacyInputIterator.


Was used to replace deprecated 

https://en.cppreference.com/w/cpp/string/basic_string/assign

Implicitly converts t to a string view sv as if by std::basic_string_view<CharT, Traits> sv = t;, then replaces the contents with those of sv, as if by assign(sv.data(), sv.size()).
 This overload participates in overload resolution only if std::is_convertible_v<const StringViewLike&,
                      std::basic_string_view<CharT, Traits>> is true and std::is_convertible_v<const StringViewLike&, const CharT*> is false.

File: src/dpp/discordvoiceclient.cpp [hunk 3]
```diff
@@ -737,43 +736,44 @@ void discord_voice_client::read_ready()
 
 		voice_payload vp{0, // seq, populate later
 		                 0, // timestamp, populate later
-		                 std::make_unique<voice_receive_t>(nullptr, std::string((char*)buffer, r))};
+		                 std::make_unique<voice_receive_t>(nullptr, std::string((char*)buffer, packet_size))};
```
[the remark this answers is not in the document]
Comment (ruslan-ilesik) [reply src/dpp/discordvoiceclient.cpp:739]: R variable was renamed to have more meaningful name especially when now it is needed more than just to create basic_string_view

File: src/dpp/dispatcher.cpp [hunk 4]
```diff
@@ -279,11 +279,11 @@ void voice_receive_t::reassign(discord_voice_client* vc, snowflake _user_id, con
 	voice_client = vc;
 	user_id = _user_id;
 
-	audio_data.assign(pcm, length);
+	audio_data.assign(pcm, pcm + length);
```
[the remark this answers is not in the document]
Comment (ruslan-ilesik) [reply src/dpp/dispatcher.cpp:282]: Same as above

**Comment by ruslan-ilesik** [comment]
> not sure this is right please check it works

Also, additionally I ran a voice record bot we have, was working just fine (just to double check)

**Review by Jaskowicz1 — APPROVED** [review]
Looks a-okay to me!

**Review by braindigitalis — APPROVED** [review]

**Comment by braindigitalis** [comment]
can you please fix the codacy issue by moving the [[maybe_unused]] before the const in the if()

**Comment by ruslan-ilesik** [comment]
> can you please fix the codacy issue by moving the [[maybe_unused]] before the const in the if()

How can i do it? This variable was there before my changes, Should i just remove it?

**Pull request merged**